File: Sesion 10/Actividades/Actividad 3/Actividad3_S10.java
import java.util.ArrayList;

class Ruta<T> {
    private static final int MAX_PARADAS = 10;
    private String nombre;
    private ArrayList<T> paradas;
    private int numParadas;

    public Ruta(String nombre) {
        this.nombre = nombre;
        this.paradas = new ArrayList<>();
        this.numParadas = 0;
    }

    public void agregarParada(T parada) {
        if (numParadas < MAX_PARADAS) {
            paradas.add(parada);
            numParadas++;
        } else {
            System.out.println("No se pueden agregar más paradas a la ruta " + nombre);
        }
    }

    public void mostrarRuta() {
        System.out.println("Ruta: " + nombre);
        for (T parada : paradas) {
            System.out.println("Parada: " + parada);
        }
    }

    public <V> void asignarVehiculo(V vehiculo) {
        System.out.println("Asignando vehículo " + vehiculo + " a la ruta " + nombre);
    }
}

class SistemaGestionRutas<T extends Ruta<?>> {
    private static final int MAX_RUTAS = 10;
    private ArrayList<T> rutas;
    private int numRutas;

    public SistemaGestionRutas() {
        this.rutas = new ArrayList<>();
        this.numRutas = 0;
    }

    public void agregarRuta(T ruta) {
        if (numRutas < MAX_RUTAS) {
            rutas.add(ruta);
            numRutas++;
        } else {
            System.out.println("No se pueden agregar más rutas al sistema");
        }
    }

    public void mostrarRutas() {
        for (T ruta : rutas) {
            ruta.mostrarRuta();
        }
    }

    public <V> void optimizarRecursos(V recurso) {
        System.out.println("Optimizando recurso: " + recurso);
    }

    public void generarInforme() {
        System.out.println("Generando informe del sistema de gestión de rutas...");
        mostrarRutas();
    }
}

public class Main {
    public static void main(String[] args) {
        Ruta<String> ruta1 = new Ruta<>("Ruta 1");
        ruta1.agregarParada("Parada A");
        ruta1.agregarParada("Parada B");

        Ruta<String> ruta2 = new Ruta<>("Ruta 2");
        ruta2.agregarParada("Parada X");
        ruta2.agregarParada("Parada Y");

        SistemaGestionRutas<Ruta<String>> sistema = new SistemaGestionRutas<>();
        sistema.agregarRuta(ruta1);
        sistema.agregarRuta(ruta2);

        sistema.mostrarRutas();
        sistema.optimizarRecursos("Camión 123");
        sistema.generarInforme();
    }
}

File: Sesion 10/Actividades/Actividad 3/Actividad3_S10.py
class Ruta:
    MAX_PARADAS = 10

    def __init__(self, nombre):
        self.nombre = nombre
        self.paradas = []
        self.num_paradas = 0

    def agregar_parada(self, parada):
        if self.num_paradas < self.MAX_PARADAS:
            self.paradas.append(parada)
            self.num_paradas += 1
        else:
            print(f"No se pueden agregar más paradas a la ruta {self.nombre}")

    def mostrar_ruta(self):
        print(f"Ruta: {self.nombre}")
        for parada in self.paradas:
            print(f"Parada: {parada}")

    def asignar_vehiculo(self, vehiculo):
        print(f"Asignando vehículo {vehiculo} a la ruta {self.nombre}")


class SistemaGestionRutas:
    MAX_RUTAS = 10

    def __init__(self):
        self.rutas = []
        self.num_rutas = 0

    def agregar_ruta(self, ruta):
        if self.num_rutas < self.MAX_RUTAS:
            self.rutas.append(ruta)
            self.num_rutas += 1
        else:
            print("No se pueden agregar más rutas al sistema")

    def mostrar_rutas(self):
        for ruta in self.rutas:
            ruta.mostrar_ruta()

    def optimizar_recursos(self, recurso):
        print(f"Optimizando recurso: {recurso}")

    def generar_informe(self):
        print("Generando informe del sistema de gestión de rutas...")
        self.mostrar_rutas()


ruta1 = Ruta("Ruta 1")
ruta1.agregar_parada("Parada A")
ruta1.agregar_parada("Parada B")

ruta2 = Ruta("Ruta 2")
ruta2.agregar_parada("Parada X")
ruta2.agregar_parada("Parada Y")

sistema = SistemaGestionRutas()
sistema.agregar_ruta(ruta1)
sistema.agregar_ruta(ruta2)

sistema.mostrar_rutas()
sistema.optimizar_recursos("Camión 123")
sistema.generar_informe()


